lib/pure/math.nim

```diff
@@ -915,27 +926,53 @@ func euclMod*[T: SomeNumber](x, y: T): T {.since: (1, 5, 1).} =
   if result < 0:
     result += abs(y)
 
-when not defined(js):
-  func c_frexp*(x: float32, exponent: var int32): float32 {.
-      importc: "frexp", header: "<math.h>".}
-  func c_frexp*(x: float64, exponent: var int32): float64 {.
-      importc: "frexp", header: "<math.h>".}
-  func frexp*[T, U](x: T, exponent: var U): T =
-    ## Split a number into mantissa and exponent.
-    ##
-    ## `frexp` calculates the mantissa `m` (a float greater than or equal to 0.5
-    ## and less than 1) and the integer value `n` such that `x` (the original
-    ## float value) equals `m * 2**n`. frexp stores `n` in `exponent` and returns
-    ## `m`.
-    runnableExamples:
-      var x: int
-      doAssert frexp(5.0, x) == 0.625
-      doAssert x == 3
+func frexp[T: float32|float64](x: T): tuple[frac: T, exp: int] {.inline.} =
+  ## Splits a number into mantissa and exponent.
+  ## The mathematical formula for this method is: `x = frac * 2 ^ exp`
+  runnableExamples:
+    doAssert frexp(8.0) == (0.5, 4)
+    doAssert frexp(0.0) == (0.0, 0)
+  when not defined(js):
+    var exp: cint
+    result.frac = c_frexp2(x, exp)
+    result.exp = exp
+  else:
+    if x == 0.0:
+      result = (0.0, 0)
+    elif x < 0.0:
+      result = frexp(-x)
+      result.frac = -result.frac
+    else:
+      var ex = trunc(log2(x))
+      result.exp = int(ex)
+      result.frac = x / pow(2.0, ex)
+      if abs(result.frac) >= 1:
+        inc(result.exp)
+        result.frac = result.frac / 2
+      if result.exp == 1024 and result.frac == 0.0:
+        result.frac = 0.99999999999999988898
+
+since (1, 5, 1):
+  export frexp
+
+func frexp*[T: float32|float64](x: T, exponent: var int): T {.inline.} =
+  ## Splits a number into mantissa and exponent.
+  ##
+  ## `frexp` calculates the mantissa m (a float greater than or equal to 0.5
```

Comment:
```suggestion
  ## `frexp` calculates the mantissa m (a float with an absolute value greater than or equal to 0.5
```

Reply:
Like I said in https://github.com/nim-lang/Nim/pull/16725#discussion_r568294568, we shouldn't duplicate the doc comment, this should be:
```nim
## Overload of `frexp` that calls `(result, exponent) = frexp(x)`.
```

and the "preferred" frexp should contain the details.